### DISCUSSION TITLE
Want to contribute to CalcUS

### DISCUSSION THREAD

**Discussion by 0xClint** [cyllab/CalcUS#21] [Developer support]
Hello developers, I want to contribute in this repo, can any one tell me how can I contribute?

**Comment by RaphaelRobidas** [comment]
Hello, thanks for your interest in the project, let me help you get started.

Firstly, you will need to fork this repository, clone your fork locally and configure CalcUS. You can follow the [installation instructions from the documentation](https://calcus.readthedocs.io/en/latest/installation.html). Make sure you can also get CalcUS running from source (see the last section of the installation page). If you get stuck, let me know and we'll debug the issue.

See also the [developer guide](https://calcus.readthedocs.io/en/latest/developer-guide.html). You should run some tests at least once before getting started just to make sure that the tests work as expected.

Afterwards, you will be ready to start coding! Did you have a specific feature or part of the platform (frontend, backend...) that you wanted to work on? I have a large list of tasks and ideas of all kinds, I can propose some improvements that would be helpful if you want.

**Comment by 0xClint** [comment]
Thank you @RaphaelRobidas , I will set CalcUS locally and let you know

**Comment by 0xClint** [comment]
@RaphaelRobidas, while building from source " docker-compose -f dev-compose.yml -f docker-compose.override.yml build " getting this error!
![Screenshot from 2023-01-29 17-09-33](https://user-images.githubusercontent.com/97947439/215323584-d7b4f1ac-076c-48c3-bc1a-d6ba5d0b5baa.png)

**Reply by RaphaelRobidas** [reply]
It seems like `pip` had trouble downloading the packages from `files.pythonhosted.org`. If you have a slow or fluctuating internet connection, you can try again a couple times and see if you get a different outcome.

Other (relatively unlikely) possible causes:
- The access to `pythonhosted.org` is being blocked by your router for some reason
- `pythonhosted.org` is not working properly

**Comment by 0xClint** [comment]
Yes it was my college's internet problem.
Thank you, the local project setup is completed.
 
While testing, when I start development environment `./start_test.sh` 
getting and error:
```
Removing network calcus_default
WARNING: Network calcus_default not found.
ERROR: The Compose file is invalid because:
Service slurm has neither an image nor a build context specified. At least one must be provided.
```

Relate to contribution, I would like to contribute more in frontend part of the project!

**Reply by RaphaelRobidas** [reply]
This is strange, it seems like Docker is trying to run the `slurm` service, which I removed a while ago from the tests. Can you check that you have the most recent `.yml` files? Both those on `master` and those on `develop` should be fine. Let me know if that solves it!

I will also look at the pending frontend tasks and suggest some appropriate ones shortly.

**Comment by 0xClint** [comment]
Yes I have checked **.yml ** files but it is creating error due to `test-compose.override.yml` and `test-compose.yml` which creates while setting environment.

**Reply by RaphaelRobidas** [reply]
Nice catch, it was indeed `test-compose.override.yml` being generated with outdated parameters. I have pushed a fix on `develop`.

Also, you should indicate "no" regarding `g16` and `orca`. These are separate quantum chemistry software packages which you won't need to contribute. CalcUS will automatically extract and install `xtb` (https://github.com/grimme-lab/xtb), a very fast and open-source quantum chemistry package. You will be able to use it to generate some data for your purposes.

**Comment by 0xClint** [comment]
Thanks, @RaphaelRobidas for helping out to set up the project locally. I also performed some tests. Now can you allot me some frontend tasks?

**Reply by RaphaelRobidas** [reply]
Here's your task: add short hash identifiers in the frontend. Specifically, in `frontend/ensemble.html` so that `/ensemble/BeLl94N7qMpDV` shows a short hash (5 characters) like the mockup below, in the top left:

![image](https://user-images.githubusercontent.com/52597883/215768244-6b534f37-a261-4bb4-a84a-09aeb156cac7.png)

Moreover, these short hashes should also show up in the ensemble map of the parent molecule page (`/molecule/<hash>`):
![image](https://user-images.githubusercontent.com/52597883/215768946-2db4814c-1f67-48eb-b5c4-7712ad4fc6eb.png)

You will have to run some calculations in order to create ensembles. You can draw any random small molecule with the sketcher and choose "Geometrical Optimization" as calculation type. This will create the ensembles as above.

Let me know if you get stuck or if you have questions. Cheers!

**Reply by 0xClint** [reply]
Just a basic doubt, How can I run the frontend dev, so that I can see the live changes in the frontend. Just like in flask we can run it by command `flask run`.

Like can you tell me some more additional commands which can help me to check the changes I have done in code?

**Reply by RaphaelRobidas** [reply]
If you modify `frontend/views.py`, all the frontend code and the page templates will be reloaded. You should see something like this in the Docker logs:
```
web_1          | [2023-01-31 15:06:04 -0500] [89] [INFO] Worker reloading: /calcus/frontend/views.py modified
web_1          | [2023-01-31 15:06:04 -0500] [89] [INFO] Worker exiting (pid: 89)
web_1          | [2023-01-31 20:06:04 +0000] [106] [INFO] Booting worker with pid: 106
```

If you modify the page templates (`frontend/templates`), the changes will *not* be automatically reloaded. However, you can trigger a refresh by resaving `frontend/views.py` without any real change. On Linux, you can use `touch frontend/views.py`.

The only case where you need to rebuild the Docker image and restart CalcUS is when you modify services other than `web`. For example, if you modify the backend code handling calculations, it will require a rebuild. In your case, you shouldn't need to do that though.

**Reply by 0xClint** [reply]
The first part of the task is done!
![Screenshot from 2023-02-01 08-45-00](https://user-images.githubusercontent.com/97947439/215936617-fa576852-0df8-4e56-8ae2-c8e6c9143c4b.png)

In the 2nd part ie. in `/molecule/hash`, we have used cytoscape-degra for branch structure css and framing. I just gone through their docs, I was not able to find any way to add more than one labels ie node name and hash in same box. 

```
var cy = (window.cy = cytoscape({
                    container: document.getElementById("ensembles_graph"),
                    boxSelectionEnabled: false,
                    autounselectify: true,
                    wheelSensitivity: 0.1,
                    layout: {
                        name: "dagre",
                    },
                    style: [
                        {
                            selector: "node",
                            style: {
                                "background-color": "data(color)",
                                shape: "round-rectangle",
                                label: "data(name)",           
                            "text-valign": "center",
                            "text-halign": "center",
                            "font-size": "10px",
                            width: "label",
                            padding: "5px",
                            "border-width": "data(bwidth)",
                            "border-color": "data(bcolor)",
                        },
                    },
                    
                    {
                        selector: "edge",
                        style: {
                            width: 4,
                                "target-arrow-shape": "triangle",
                                "line-color": "#000000",
                                "target-arrow-color": "#000000",
                                "curve-style": "bezier",
                            },
                        },
                    ],
                    elements: JSON.parse(data),
                }));
```

**Reply by RaphaelRobidas** [reply]
Great for the first part! For the second, I thought it could be possible to use a linebreak ("\n") in the label to have the hash on a second line. I haven't verified that it works though. If it doesn't, you could check if there is a way to wrap text on multiple lines if it is too long and then just add the hash at the end of the label.

**Reply by 0xClint** [reply]
I tried all the possible cases, but cytoscape-degra only take data in "data(<element)" format only or a simple string.
If we add any things data(<element>) it is not showing the element's value ie:
```
label : "data(name)" ==output=> name
label: "data(name)"+"hello" ==output => data(name)hello
label:"data(name) hello" ==output=>  data(name) hello
label : "data(name)  \n hello" ==output==> data(name) \n hello
```

It needs the data in only data(<element>) format else it will not pass the element's value
![Screenshot from 2023-02-02 16-39-10](https://user-images.githubusercontent.com/97947439/216308982-7a806058-47eb-4c61-9648-00bd84f3c38d.png)
![Screenshot from 2023-02-02 16-38-52](https://user-images.githubusercontent.com/97947439/216308975-14a15ddd-0a3f-4d6c-adc2-ba90f337e510.png)
![Uploading Screenshot from 2023-02-02 16-39-10.png…]()

**Reply by 0xClint** [reply]
I tried all the possible cases, but cytoscape-degra only take data in "data(<element)" format only or a simple string.
If we add any things data(<element>) it is not showing the element's value ie:
```
label : "data(name)" ==output=> Drawn structure
label: "data(name)"+"hello" ==output => data(name)hello
label:"data(name) hello" ==output=>  data(name) hello
label : "data(name)  \n hello" ==output==> data(name) \n hello
```

It needs the data in only data(<element>) format else it will not pass the element's value
![Screenshot from 2023-02-02 16-39-10](https://user-images.githubusercontent.com/97947439/216308982-7a806058-47eb-4c61-9648-00bd84f3c38d.png)
![Screenshot from 2023-02-02 16-38-52](https://user-images.githubusercontent.com/97947439/216308975-14a15ddd-0a3f-4d6c-adc2-ba90f337e510.png)
![Screenshot from 2023-02-02 16-39-10](https://user-images.githubusercontent.com/97947439/216309011-6860a0c9-b1b1-49f8-bd6e-de12902a7a24.png)

**Reply by RaphaelRobidas** [reply]
I see. You will have to modify the `name` property then to contain the short hash already.

If the newline doesn't work properly, just add the short hash in parentheses.

**Reply by 0xClint** [reply]
If we want to do this we will need to modify the name property of data object. before passing `data `to `var cy` I will need to modify its `name` element, we can't directly add it to data by just adding it to parentheses

```
    function refresh_map() {
        $.ajax({
            method: "GET",
            url: "/ensemble_map/{{ molecule.id }}",
            headers: {
                "X-CSRFToken": "{{ csrf_token }}",
            },
            success: function (data) {
                var id = '{{molecule.id}}'
                console.log('{{molecule.id}}'+ data)
           
                var cy = (window.cy = cytoscape({
                    container: document.getElementById("ensembles_graph"),
                    boxSelectionEnabled: false,
                    autounselectify: true,
                    wheelSensitivity: 0.1,
                    layout: {
                        name: "dagre",
                    },
                    style: [
                        {
                            selector: "node",
                            style: {
                                "background-color": "data(color)",
                                shape: "round-rectangle",
                                label: "data({{molecule.id}})",

                            "text-valign": "center",
                            "text-halign": "center",
                            "font-size": "10px",
                            width: "label",
                            padding: "5px",
                            "border-width": "data(bwidth)",
                            "border-color": "data(bcolor)",
                        },
                    },
                    
                    {
                        selector: "edge",
                        style: {
                            width: 4,
                                "target-arrow-shape": "triangle",
                                "line-color": "#000000",
                                "target-arrow-color": "#000000",
                                "curve-style": "bezier",
                            },
                        },
                    ],
                    elements: JSON.parse(data),
                }));
                cy.on("tap", "node", function () {
                    //try {
                    //	window.open( this.data('href') );
                    //} catch(e){
                    window.location.href = this.data("href");
                    //}
                });
            },
        });
    }
```

**Reply by RaphaelRobidas** [reply]
It would be easier to modify the `ensemble_map` function in `frontend/views.py`

**Reply by 0xClint** [reply]
I modified the ensemble_map function by adding one more element in object `label`
```
label="{} ({})".format(e.name,mol.id)
        nodes += """{{ "data": {{"label":"{}", "id": "{}", "name": "{}", "href": "/ensemble/{}", "color": "{}"{}}} }},""".format(
           label, e.id, e.name, e.id, e.get_node_color, border_text
```
![Screenshot from 2023-02-04 14-49-05](https://user-images.githubusercontent.com/97947439/216759391-8443b67c-5fd3-4324-b61e-df8a3f8ff3e9.png)

**Reply by RaphaelRobidas** [reply]
That's great. Could you just adjust the label to have the first 5 characters of the id? After that, you could send a PR to the develop branch!

**Reply by 0xClint** [reply]
The changes are done!
![Screenshot from 2023-02-04 21-27-58](https://user-images.githubusercontent.com/97947439/216777180-55df6b40-3705-4714-9e1c-63ccd29caa66.png)
I have also created a PR to develop branch

**Comment by 0xClint** [comment]
Can you allot me some more tasks as an issue?

**Reply by RaphaelRobidas** [reply]
I have added two frontend improvements as Github issues. You can check them out and work on that, if you'd like!

**Reply by 0xClint** [reply]
Yes, I will start working on it!

**Comment by 0xClint** [comment]
Hey @RaphaelRobidas, can you suggest me some reference material for Django, because I want to contribute more to this project :)

**Reply by RaphaelRobidas** [reply]
You can check out the general Django tutorial to give you an overall picture of the application: https://docs.djangoproject.com/en/4.1/intro/tutorial01/

Some more relevant links around the frontend:
- https://docs.djangoproject.com/en/4.1/topics/templates/
- https://docs.djangoproject.com/en/4.1/topics/http/urls/
- https://docs.djangoproject.com/en/4.1/topics/http/views/

**Comment by 0xClint** [comment]
@RaphaelRobidas I found that project CalcUS is in Google Summer of Code under OpenChemistry Organization. I would like to participate in the Google Summer of Code program this year by contributing in CalcUS and I am seeking guidance on project ideas.

I am excited about the opportunity to work on this project that aligns with my interests and skills, and I am eager to work with you as a mentee. However, I am unsure about where to start in terms of identifying potential project idea.

I was hoping that you could provide me with some guidance and advice on how to propose to project ideas, and perhaps offer some insight into what qualities make a successful GSoC project. I am also interested in learning more about what you look for in a potential GSoC participant and how I can make myself stand out as a strong candidate.

I look forward to hearing from you soon.

**Reply by RaphaelRobidas** [reply]
Hello @Omkar0803,

There aren't any big frontend-only projects planned at the moment. There are some improvements and new features that could be made (like the ability to search through the results), but they don't quite have a common theme. There are much more fullstack and backend project ideas, which might not be reasonable for you if you don't have much experience with Django.

Are you quite knowledgeable in CSS, Javascript and JQuery? If so, it could make sense to bundle the different improvements, fixes and factoring into a project about "Improving the web frontend of the platform". A mastery of these languages is preferrable in order to improve the code quality by a good margin and tackle relatively advanced challenges. If that's not exactly your case, you might want to check out other GSoC possibilities for this year.

Thanks for the interest and contributions to the project!

**Comment by 0xClint** [comment]
Thank you for your response regarding project ideas for the Google Summer of Code program. I appreciate your willingness to guide me in identifying potential projects. Yes, I am having quite enough experience in HTML, CSS, Javascript and JQuery. I have built lot of [Projects](https://www.omkardarde.me/) in this tech stack.

I would love to hear your ideas and suggestions for ways that I can contribute to the project and improve the web frontend of the platform. I am eager to learn from your experience and expertise and to collaborate with you to ensure that my work aligns with the goals of the CalcUS project.

I am particularly interested in hearing your thoughts on:
    Areas of the web frontend that need improvement
    New features or tools that could be added to the platform
    Best practices for web development and user experience design

I am committed to making a meaningful contribution to the CalcUS project and to ensuring that my work has a positive impact on the platform and its users. I appreciate your time and guidance as I embark on this exciting journey.

Thank you for your support!

**Reply by RaphaelRobidas** [reply]
Here are some rough notes of improvements I have been thinking about:
- There should be a search functionality to find relevant calculations. That would imply a portion of backend code to perform the actual search query.
- The warnings that are sometimes displayed on the launch page could be extended into a complete "best practices" module, which may show more detailed warnings with explanations based on the parameters that were entered. There could be a warning symbol near the parameter which triggered the warning, and clicking on that symbol would open a popup which displays further information.
- The web spreadsheet on CalcUS has most of the basic features we need, but we can't enter a reference to a cell simply by clicking on the cell (like in Excel). The functionality is included only in a paid version of the JS plugin. I'm still looking for another way to get this feature.

**Reply by 0xClint** [reply]
Thank you @RaphaelRobidas
